### PR TITLE
Handle Etcd connection error

### PR DESCRIPTION
**Pull request by ntk148v**
Close: #48

### CODE REVIEW

File: pkg/autohealer/manager.go [hunk 1]
```diff
@@ -135,32 +138,41 @@ func (hm *Manager) stopWorker(name string) {
...
 	level.Info(hm.logger).Log("msg", "Autohealer manager is stopped!")
 }
 
 func (hm *Manager) save() {
 	for e := range hm.rqt.Iter() {
 		hm.wg.Add(1)
 		go func(e utils.RegistryItem) {
-			defer hm.wg.Done()
+			defer func() {
+				fmt.Println(e.Name)
```
Comment (vtdat) [inline pkg/autohealer/manager.go]: should not print here.

File: pkg/autoscaler/manager.go [hunk 2]
```diff
@@ -67,11 +70,17 @@ func (m *Manager) Reload() {
...
+	if m.state == model.StateStopping || m.state == model.StateStopped {
+		return
+	}
 	level.Info(m.logger).Log("msg", "Stopping autoscale manager...")
+	m.state = model.StateStopping
 	close(m.stop)
 	m.save()
 	// Wait until all scalers shut down
 	m.wg.Wait()
+	m.state = model.StateStopped
```
Comment (vtdat) [inline pkg/autoscaler/manager.go]: we do not check any further with `StateStopped`. Is this really necessary here?

for me `StateStopping` is good enough.

File: pkg/cluster/cluster.go [hunk 3]
```diff
@@ -129,7 +131,7 @@ func New(cid, bindAddr string, l log.Logger, e *etcdv3.Client) (*Cluster, error)
 	}
 
 	// Grant lease
-	leaseResp, err := c.etcdcli.Grant(context.Background(), DefaultLeaseTTL)
+	leaseResp, err := c.etcdcli.DoGrant(nil, DefaultLeaseTTL)
```
Comment (vtdat) [inline pkg/cluster/cluster.go]: why `nil`?

File: cmd/faythe/main.go [hunk 4]
```diff
@@ -93,7 +95,7 @@ func main() {
 
 	var (
 		etcdConf = etcdv3.Config{}
-		etcdCli  = &etcdv3.Client{}
+		etcdCli  = &etcd.V3{}
```
Comment (vtdat) [inline cmd/faythe/main.go]: please use `etcdcli` for unification.

File: cmd/faythe/main.go [hunk 5]
```diff
@@ -139,15 +141,20 @@ func main() {
...
+	stopFunc := func() {
 		watchCancel()
 		fas.Stop()
 		cls.Stop()
-		nr.Stop()
+		fah.Stop()
 		etcdCli.Close()
-		level.Info(logger).Log("msg", "Faythe is stopped, bye bye!")
+	}
+	defer func() {
```
Comment (vtdat) [inline cmd/faythe/main.go]: in which cases this can be triggered?

File: config/config.go [hunk 6]
```diff
@@ -103,11 +104,16 @@ type EtcdConfig struct {
 
 	// PermitWithoutStream when set will allow client to send keepalive pings to server without any active streams(RPCs).
 	PermitWithoutStream bool `yaml:"permit_without_stream,omitempty"`
+
+	// DialOptions is a list of dial options for the grpc client (e.g., for interceptors).
+	// For example, pass "grpc.WithBlock()" to block until the underlying connection is up.
+	// Without this, Dial returns immediately and connecting the server happens in background.
+	DialOptions []grpc.DialOption
```
Comment (vtdat) [inline config/config.go]: need update docs or config example.

we shouldn't force users to read our codes for documentation of `options`

File: pkg/autohealer/healer.go [hunk 7]
```diff
@@ -115,7 +89,7 @@ func (h *Healer) run(ctx context.Context, wg *sync.WaitGroup, nc chan map[string
 				if err != nil {
 					level.Error(h.logger).Log("msg", "Executing query failed, skip current interval",
 						"query", h.Query, "err", err)
-					h.state = stateFailed
+					h.state = model.StateFailed
```
Comment (vtdat) [inline pkg/autohealer/healer.go]: don't we have our states checked? 

or this can be used `in future`?

File: pkg/autohealer/manager.go [hunk 8]
```diff
@@ -62,6 +64,7 @@ func NewManager(l log.Logger, e *etcdv3.Client, c *cluster.Cluster) *Manager {
 		cluster: c,
 	}
 	hm.load()
+	hm.state = model.StateActive
```
Comment (vtdat) [inline pkg/autohealer/manager.go]: same here

File: pkg/autoscaler/manager.go [hunk 9]
```diff
@@ -183,27 +192,29 @@ func (m *Manager) save() {
 	for i := range m.rgt.Iter() {
 		m.wg.Add(1)
 		go func(i RegistryItem) {
-			defer m.wg.Done()
+			defer func() {
+				m.stopScaler(i.Key)
+				m.wg.Done()
+			}()
```
Comment (vtdat) [inline pkg/autoscaler/manager.go]: should pass `i.Key` param inside to avoid unwanted misbehaviours.

File: pkg/autohealer/manager.go [hunk 10]
```diff
@@ -135,32 +138,41 @@ func (hm *Manager) stopWorker(name string) {
...
 func (hm *Manager) save() {
 	for e := range hm.rqt.Iter() {
 		hm.wg.Add(1)
 		go func(e utils.RegistryItem) {
-			defer hm.wg.Done()
+			defer func() {
+				fmt.Println(e.Name)
+				hm.stopWorker(e.Name)
+				hm.wg.Done()
+			}()
```
Comment (vtdat) [inline pkg/autohealer/manager.go]: should pass `e.Name` param inside to avoid unwanted misbehaviours.

File: pkg/autohealer/manager.go (hunk 1 above)
Comment (ntk148v) [reply pkg/autohealer/manager.go]: ok...

File: pkg/cluster/cluster.go (hunk 3 above)
Comment (ntk148v) [reply pkg/cluster/cluster.go]: if `nil`, `context.Background()` will be used. Sometimes we pass request's context, sometimes we don't. Golang doesn't accept a default nil value like Python, so...

https://github.com/vCloud-DFTBA/faythe/blob/900c9ae072f2cc29325985fdb2171a1a4ae1482b/pkg/etcd/v3.go#L46

File: cmd/faythe/main.go (hunk 4 above)
Comment (ntk148v) [reply cmd/faythe/main.go]: ok.

File: cmd/faythe/main.go (hunk 5 above)
Comment (ntk148v) [reply cmd/faythe/main.go]: At the end when Faythe received a TERM signal.

File: config/config.go (hunk 6 above)
Comment (ntk148v) [reply config/config.go]: This option is transparent to users, no yaml serialization :ok_hand: 

https://github.com/etcd-io/etcd/blob/master/clientv3/config.go#L70

File: pkg/autohealer/healer.go (hunk 7 above)
Comment (ntk148v) [reply pkg/autohealer/healer.go]: For future use.

File: pkg/autohealer/manager.go (hunk 8 above)
Comment (ntk148v) [reply pkg/autohealer/manager.go]: Ditto.

File: pkg/autohealer/manager.go (hunk 10 above)
Comment (ntk148v) [reply pkg/autohealer/manager.go]: Unnecessary

File: pkg/autoscaler/manager.go (hunk 9 above)
Comment (ntk148v) [reply pkg/autoscaler/manager.go]: unnecessary

File: pkg/cluster/cluster.go (hunk 3 above)
Comment (vtdat) [reply pkg/cluster/cluster.go]: pass `context.Background()` then please, for better reading experience :D

File: pkg/autoscaler/manager.go (hunk 2 above)
Comment (vtdat) [reply pkg/autoscaler/manager.go]: for future purpose. resolved.

File: api/api.go [hunk 11]
```diff
@@ -16,14 +16,14 @@ package api
 
 import (
 	"encoding/json"
+	"github.com/vCloud-DFTBA/faythe/pkg/common"
```
Comment (vtdat) [inline api/api.go]: import positioning

File: cmd/faythe/main.go [hunk 12]
```diff
@@ -111,44 +111,50 @@ func main() {
 
 	// Init Etcdv3 client
 	copier.Copy(&etcdConf, config.Get().EtcdConfig)
-	etcdCli, err = etcdv3.New(etcdConf)
+	etcdcli, err = common.New(etcdConf)
```
Comment (vtdat) [inline cmd/faythe/main.go]: I'm abit confused. `common.New()` doesn't make sense for me. 

Ofc, I understand but others might: "initializing a `common` instance?"

Could it be changed to smt like `common.NewEtcdCli()`?

File: pkg/autohealer/healer.go [hunk 13]
```diff
@@ -70,7 +44,7 @@ type Healer struct {
 	done       chan struct{}
 	logger     log.Logger
 	mtx        sync.RWMutex
-	state      healerState
+	state      common.State
```
Comment (vtdat) [inline pkg/autohealer/healer.go]: as this will be persistent, move to `model` package?

Comment (vtdat) [reply pkg/autohealer/healer.go]: `State` should be in `model` pkg. I strongly argue that if you have second thought.

File: pkg/autoscaler/manager.go [hunk 14]
```diff
@@ -182,43 +190,52 @@ func (m *Manager) getBackend(key string) (metrics.Backend, error) {
...
+			switch it := i.Value.(type) {
+			case *Scaler:
+				it.Alert = &it.alert.State
+			default:
+				level.Error(m.logger).Log("msg", "Registry can contains only Scalers",
+					"name", i.Name)
+				return
+			}
```
Comment (vtdat) [inline pkg/autoscaler/manager.go]: idk how this work but this should be present at https://github.com/vCloud-DFTBA/faythe/pull/58/commits/a847a34956f34274ad8290fdcffb1956cb54f3e6#diff-74a7d869ea73be210fc3893445d49df0R187-R199

File: pkg/autoscaler/manager.go [hunk 15]
```diff
@@ -229,32 +246,39 @@ func (m *Manager) rebalance() {
...
 				if ok2 {
-					scaler.Alert = &scaler.alert.State
+					switch s := scaler.(type) {
+					case *Scaler:
+						s.Alert = &s.alert.State
+					default:
+						level.Error(m.logger).Log("msg", "Registry can contains only Scalers",
+							"name", name)
+						return
+					}
```
Comment (vtdat) [inline pkg/autoscaler/manager.go]: same here

**Pull request merged**